File: Project/C++/src/classification.cpp
#include <vector>
#include <algorithm>
using namespace std;

bool findRule(vector<rule> v, rule r)
{
    for (int i = 0; i < v.size(); i++)
    {
        if (v[i].compareTo(r) == 0)
        {
            return true;
        }
    }
    return false;
}

class classification
{
private:
    int type;
    vector<rule> rules;

public:
    classification(int c)
    {
        type = c;
    }

    void addRule(rule r)
    {

        if (!findRule(rules, r))
        {
            rules.push_back(r);
        }
    }

    int getClassification()
    {
        return type;
    }

    bool matches(transaction t)
    {
        for (int i = 0; i < rules.size(); i++)
        {
            if (rules[i].matches(t))
            {
                return true;
            }
        }
        return false;
    }

    string toString()
    {
        string res = "";
        res.append("Classification: " + to_string(type) + "\n");
        for (int i = 0; i < rules.size(); i++)
        {
            res.append(rules[i].toString() + "\n");
        }
        return res;
    }
};

File: Project/Java/Classification.java
import java.util.ArrayList;

public class Classification {

  private int classification;
  private ArrayList<Rule> rules;

  public Classification(int classification) {
    this.classification = classification;
    rules = new ArrayList<>();
  }

  public void addRule(Rule r) {
    if (!rules.contains(r)) {
      rules.add(r);
    }
  }

  public int getClassification() {
    return classification;
  }

  public boolean matches(Transaction t) {
    //		System.out.println("-----------------------------------------------");
    //		System.out.println("Class: " + classification);
    for (Rule rule : rules) {
      //			System.out.println(rule.toString());
      //			System.out.println(t.toString());
      if (rule.matches(t)) {
        //				System.out.println("Accept");
        //				System.out.println("-----------------------------------------------");
        return true;
      }
    }
    //		System.out.println("Reject");
    //		System.out.println("-----------------------------------------------");
    return false;
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    sb.append("Classification: " + classification + "\n");
    for (Rule r : rules) {
      sb.append(r.toString() + "\n");
    }
    return sb.toString();
  }
}
